lib/Parse/ParseType.cpp

```diff
@@ -520,7 +520,16 @@ ParserResult<TypeRepr> Parser::parseDeclResultType(Diag<> MessageID) {
     consumeToken(tok::code_complete);
     return makeParserCodeCompletionStatus();
   }
-  return parseType(MessageID);
+
+  auto result = parseType(MessageID);
+
+  if (result.isNonNull() && Tok.is(tok::r_square)) {
+    auto diag = diagnose(Tok, diag::extra_rbracket);
+    diag.fixItInsert(result.get()->getStartLoc(), getTokenText(tok::l_square));
+    consumeToken();
+    return makeParserErrorResult(new (Context) ErrorTypeRepr(Tok.getLoc()));
+  }
+  return result;
```

Comment:
I don't think this implementation handles type in *non* decl result position, right? Like param (e.g. `func foo(x: Int])`) or `where` clause (e.g. `where T == Int]`).
Have you tried to do this in `parseType()` instead of `parseDeclResultType()`?

Reply:
Correct. Putting it in `parseType()` is a much more invasive change, because a trailing `]` would be fine in many cases, the most obvious of which is arrays and dictionaries themselves (`parseTypeCollection()` calling `parseType()` to get the element type). 

One solution is an additional boolean arg passed to `parseType()` for whether or not to leave a trailing `]` alone or greedily complain. I'm happy to do that, but I'm not sure whether this mistake is common enough to be worth the complexity of passing that context around through the different `parseTypeFoo()` functions, so I started here.

Reply:
Actually, worse than that, parsing types in positions where they are at the end of subscript expressions would be a big problem (e.g. `self[key.stringValue as! Key] = value` -- from Codable).

Reply:
Agreed. Thanks for explanation :)